recipes-extended/xen/files/libxl-blktap3-support.patch

```diff
@@ -0,0 +1,703 @@
++ *
++ * This program is distributed in the hope that it will be useful,
++ * but WITHOUT ANY WARRANTY; without even the implied warranty of
++ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
++ * GNU Lesser General Public License for more details.
++ */
++
++#include "libxl_osdeps.h" /* must come before any other headers */
++#include "libxl_internal.h"
++
++#include <tap-ctl.h>
++
++#include "list.h"   /* include for list_head structure */
++
++static int blktap_find(const char *type, const char *path, struct tap_list *tap)
++{
++    struct list_head list; /* Note: structure name updated */
++    struct tap_list *entry, *next_t;
++    int ret = -ENOENT, err;
++
++    /* TAILQ_INIT(&list);--> old function */
++    INIT_LIST_HEAD(&list);
++
++    if ((err = tap_ctl_list(&list)))
++        return err;
++
++    /* TAILQ_EMPTY(&list)--> old function */
++    if (list_empty(&list))
++        return ret;
++
++    tap_list_for_each_entry_safe(entry, next_t, &list) {
++
++        if (type && (!entry->type || strcmp(entry->type, type)))
++            continue;
++
++        if (path && (!entry->path || strcmp(entry->path, path)))
++            continue;
++
++        *tap = *entry;
++	tap->type = tap->path = NULL;
++        ret = 0;
++        break;
++    }
++
++    tap_ctl_list_free(&list);
++
++    return ret;
++}
++
++/**
++ * blktap3 doesn't require blkback, so it's always available.
++ */
++int libxl__blktap_enabled(libxl__gc *gc)
++{
++    return 1;
++}
++
++char *libxl__blktap_devpath(libxl__gc *gc, const char *disk,
++		libxl_disk_format format)
++{
++    const char *type = NULL;
++    char *params, *devname = NULL;
++    struct tap_list tap;
++    int err = 0;
++    int flags = 0;
++
++    type = libxl__device_disk_string_of_format(format);
++    if (!(err = blktap_find(type, disk, &tap))) {
++        LOG(DEBUG, "found tapdisk\n");
++	devname = libxl__sprintf(gc, "/dev/xen/blktap-2/tapdev%d", tap.minor);
++        if (devname)
++            return devname;
++		/* return 0; revert-back-changes*/
++    }
++
++    LOG(DEBUG, "tapdisk not found\n");
++
++	/* TODO Should we worry about return codes other than ENOENT? */
++
++    params = libxl__sprintf(gc, "%s:%s", type, disk);
++
++    /* tap_ctl_create(params, 0, -1, NULL) --> old function call  */
++    if (!(err = tap_ctl_create(params, &devname, flags, -1, 0, 0))) {
++        LOG(DEBUG, "created tapdisk\n");
++        return devname; /*revert-back-changes*/
++    }
++
++    LOG(ERROR, "error creating tapdisk: %s\n", strerror(err));
++
++    return NULL;/*revert-back-changes*/
++}
++
++int libxl__device_destroy_tapdisk(libxl__gc *gc, const char *be_path)
++{
++    char *disk;
++    int err;
++    struct tap_list tap;
++
++	LOG(DEBUG, "destroying tapdisk %s\n", be_path);
++
++    disk = strchr(be_path, ':');
++    if (!disk) {
++        LOG(ERROR, "Unable to parse params %s", be_path);
++        return ERROR_INVAL;
++    }
++
++    *disk++ = '\0';
```

Comment:
This seems really hackish to me.  I would rather see ```strncpy``` from 0 to the element pointed to by ```strchr``` into a new local variable that is the exact size we want (plus room for the '\0').  

Perhaps more importantly, 
```
disk = strchr(be_path, ';'); 
``` 
returns a pointer into ```be_path```, which you're then modifying with
```
*disk++ = '\0';
```
be_path is declared as const but it's just the pointer to first element of the char array that's const.  You're offset into the char* so const won't apply, and thus you're modifying the original be_path.  Doesn't look like that's supposed to be the intended behaviour of the function so in my mind, this absolutely needs to be fixed.

Reply:
Yes, this needs to be fixed, I will do so by copying the element into a new local variable. As earlier this change was retained from the base libxl patch (https://github.com/qiaohaiming/blktap3/blob/master/libxl.diff)